### PR TITLE
pkgs/build-support/trivial-builders: add runCommandLocal

### PR DESCRIPTION
**Pull request by Profpatsch**
A definition I’ve been copy-pasting everywhere so far, so it’s finally
time to add it to nixpkgs.

I’m using a remote builder for my regular nix builds, so trivial
`runCommand`s which first try a substitution and then copy the inputs
to the builder to run for 0.2s are quite noticable.

If we just always build these, we gain some build time, so let’s make
it easy to switch from remote to local.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Infinisil

### CODE REVIEW

File: pkgs/build-support/trivial-builders.nix [hunk 1]
```diff
@@ -21,10 +27,26 @@ rec {
...
   runCommand = runCommandNoCC;
-  runCommandNoCC = runCommand' stdenvNoCC;
-  runCommandCC = runCommand' stdenv;
+  runCommandLocal = runCommandNoCCLocal;
+
+  runCommandNoCC = runCommand' false stdenvNoCC;
+  runCommandNoCCLocal = runCommand' true stdenvNoCC;
+
+  runCommandCC = runCommand' false stdenv;
+  runCommandCCLocal = runCommand' true stdenv;
```
Comment (grahamc) [inline pkgs/build-support/trivial-builders.nix]: Let's get these documented: https://nixos.org/nixpkgs/manual/#chap-trivial-builders

Comment (Profpatsch) [reply pkgs/build-support/trivial-builders.nix]: It’s worrying to me that the API docs differ between the manual and the docstrings.

Comment (grahamc) [reply pkgs/build-support/trivial-builders.nix]: Go ahead and update them, too :)

Comment (Profpatsch) [reply pkgs/build-support/trivial-builders.nix]: Done.

File: pkgs/build-support/trivial-builders.nix [hunk 2]
```diff
@@ -2,11 +2,17 @@
...
     stdenv.mkDerivation ({
       inherit name buildCommand;
       passAsFile = [ "buildCommand" ];
-    } // env);
+    }
+    // env
+    // (if runLocal then {
+          preferLocalBuild = true;
+          allowSubstitutes = false;
+       } else {})
```
Comment (Infinisil) [inline pkgs/build-support/trivial-builders.nix]: I feel like this should be the other way around, such that `env` can override `preferLocalBuild`. Also could use `optionalAttrs runLocal { ... }` here

**Review by infinisil — CHANGES_REQUESTED** [review]

File: pkgs/build-support/trivial-builders.nix (hunk 2 above)
Comment (Profpatsch) [reply pkgs/build-support/trivial-builders.nix]: You are absolutely right, see force push.

Tested manually with

```
nix-build -E 'with import ./. {}; runCommandLocal "foo" { x = "y"; preferLocalBuild = false; } "echo $x"'
```

**Review by infinisil — APPROVED** [review]

**Pull request merged**